File: 2882-ways-to-express-an-integer-as-sum-of-powers/ways-to-express-an-integer-as-sum-of-powers.cpp
class Solution {
public:
    int MOD = 1e9+7;
    int numberOfWays(int n, int x) {
        int top = 0;
        while (pow(top+1, x) <= n) {
            top ++;
        }
        // cout << top;
        vector<int> f(n+1, 0);
        f[0] = 1;           // 0 way to form 0
        for(int j = 1; j <= top; j++){
            for (int i = n; i >= 1; i--){
                int d = pow(j,x);
                if (i < d) break; 
                f[i] = (f[i] + f[i-d]) % MOD;
                // cout << f[i] << " ";
            }
        }
        return f[n];
    }
};

File: 2882-ways-to-express-an-integer-as-sum-of-powers/ways-to-express-an-integer-as-sum-of-powers.py
class Solution:
    def numberOfWays(self, n: int, x: int) -> int:
        f = [0 for _ in range(n+1)]
        f[0] = 1
        MOD = 1e9+7
        t = 0
        while pow(t+1,x) <= n:
            t += 1
        
        for i in range(1,t+1):
            d = pow(i,x)
            for j in range(n, d-1, -1):
                f[j] = (f[j] + f[j-d]) % MOD;

        return int(f[n])